File: centsible-ui/app/models/rule/rule-matching.ts
import type {CategoryType} from "~/models/category/category";
import type {Rule, RuleCondition} from "~/models/rule/rule";

export interface RuleContext {
  description: string;
  amount: number;
  type: CategoryType;
  accountId: string;
}

export interface RuleEffects {
  categoryId?: number;
  tagIds: number[];
  categoryRuleName?: string;
}

export function conditionMatches(condition: RuleCondition, context: RuleContext): boolean {
  switch (condition.field) {
    case 'DESCRIPTION': {
      const value = condition.value.toLowerCase();
      const description = context.description.toLowerCase();
      switch (condition.operator) {
        case 'CONTAINS': return description.includes(value);
        case 'EQUALS': return description === value;
        case 'STARTS_WITH': return description.startsWith(value);
        default: return false;
      }
    }
    case 'AMOUNT': {
      if (condition.value.trim() === '') return false;
      const target = Number(condition.value);
      if (Number.isNaN(target)) return false;
      switch (condition.operator) {
        case 'GT': return context.amount > target;
        case 'GTE': return context.amount >= target;
        case 'LT': return context.amount < target;
        case 'LTE': return context.amount <= target;
        case 'EQUALS': return context.amount === target;
        default: return false;
      }
    }
    case 'DIRECTION':
      return condition.operator === 'IS' && context.type.toLowerCase() === condition.value.toLowerCase();
    case 'ACCOUNT':
      return condition.operator === 'IS' && context.accountId.toLowerCase() === condition.value.toLowerCase();
    default:
      return false;
  }
}

export function ruleMatches(rule: Rule, context: RuleContext): boolean {
  if (rule.conditions.length === 0) return false;
  return rule.matchAll
    ? rule.conditions.every((c) => conditionMatches(c, context))
    : rule.conditions.some((c) => conditionMatches(c, context));
}

export function evaluateRules(rules: Rule[], context: RuleContext): RuleEffects {
  let categoryId: number | undefined;
  let categoryRuleName: string | undefined;
  const tagIds: number[] = [];
  for (const rule of rules) {
    if (!rule.enabled || !ruleMatches(rule, context)) continue;
    for (const action of rule.actions) {
      if (action.type === 'SET_CATEGORY') {
        const category = action.category;
        if (categoryId === undefined && category?.id != null && category.type === context.type) {
          categoryId = category.id;
          categoryRuleName = rule.name;
        }
      } else if (action.type === 'ADD_TAG') {
        const tagId = action.tag?.id;
        if (tagId != null && !tagIds.includes(tagId)) tagIds.push(tagId);
      }
    }
  }
  return {categoryId, tagIds, categoryRuleName};
}

File: centsible-api/src/main/kotlin/beer/thierry/centsible/api/model/rule/RuleMatching.kt
package beer.thierry.centsible.api.model.rule

import beer.thierry.centsible.api.model.category.CategoryType
import java.math.BigDecimal
import java.util.UUID

data class RuleContext(
    val description: String,
    val amount: BigDecimal,
    val type: CategoryType,
    val accountId: UUID,
)

data class RuleEffects(
    val categoryId: Long? = null,
    val tagIds: Set<Long> = emptySet(),
) {
    val hasEffect: Boolean get() = categoryId != null || tagIds.isNotEmpty()
}

object RuleMatching {

    val VALID_OPERATORS: Map<RuleField, Set<RuleOperator>> = mapOf(
        RuleField.DESCRIPTION to setOf(RuleOperator.CONTAINS, RuleOperator.EQUALS, RuleOperator.STARTS_WITH),
        RuleField.AMOUNT to setOf(RuleOperator.GT, RuleOperator.GTE, RuleOperator.LT, RuleOperator.LTE, RuleOperator.EQUALS),
        RuleField.DIRECTION to setOf(RuleOperator.IS),
        RuleField.ACCOUNT to setOf(RuleOperator.IS),
    )

    /** Whether a single [condition] holds for [context]; an operator invalid for the field never matches. */
    fun conditionMatches(condition: RuleConditionDTO, context: RuleContext): Boolean = when (condition.field) {
        RuleField.DESCRIPTION -> when (condition.operator) {
            RuleOperator.CONTAINS -> context.description.contains(condition.value, ignoreCase = true)
            RuleOperator.EQUALS -> context.description.equals(condition.value, ignoreCase = true)
            RuleOperator.STARTS_WITH -> context.description.startsWith(condition.value, ignoreCase = true)
            else -> false
        }

        RuleField.AMOUNT -> {
            val target = condition.value.toBigDecimalOrNull() ?: return false
            val cmp = context.amount.compareTo(target)
            when (condition.operator) {
                RuleOperator.GT -> cmp > 0
                RuleOperator.GTE -> cmp >= 0
                RuleOperator.LT -> cmp < 0
                RuleOperator.LTE -> cmp <= 0
                RuleOperator.EQUALS -> cmp == 0
                else -> false
            }
        }

        RuleField.DIRECTION -> condition.operator == RuleOperator.IS &&
            context.type.name.equals(condition.value, ignoreCase = true)

        RuleField.ACCOUNT -> condition.operator == RuleOperator.IS &&
            context.accountId.toString().equals(condition.value, ignoreCase = true)
    }

    /** Whether [conditions] hold for [context] under [matchAll] (all vs any). No conditions never matches. */
    fun matches(matchAll: Boolean, conditions: List<RuleConditionDTO>, context: RuleContext): Boolean {
        if (conditions.isEmpty()) return false
        return if (matchAll) conditions.all { conditionMatches(it, context) }
        else conditions.any { conditionMatches(it, context) }
    }

    /** A rule matches when (matchAll) all / (else) any of its conditions hold. No conditions never matches. */
    fun ruleMatches(rule: RuleDTO, context: RuleContext): Boolean =
        matches(rule.matchAll, rule.conditions, context)

    /**
     * Evaluates [rules] (expected in priority order — highest first) against [context]. The first
     * matching rule whose SET_CATEGORY action targets a category of the transaction's own type wins
     * the category; ADD_TAG actions from every matching rule accumulate. Disabled rules are skipped.
     */
    fun evaluate(rules: List<RuleDTO>, context: RuleContext): RuleEffects {
        var categoryId: Long? = null
        val tagIds = LinkedHashSet<Long>()
        for (rule in rules) {
            if (!rule.enabled || !ruleMatches(rule, context)) continue
            for (action in rule.actions) {
                when (action.type) {
                    RuleActionType.SET_CATEGORY -> {
                        val category = action.category
                        if (categoryId == null && category?.id != null && category.type == context.type) {
                            categoryId = category.id
                        }
                    }

                    RuleActionType.ADD_TAG -> action.tag?.id?.let { tagIds.add(it) }
                }
            }
        }
        return RuleEffects(categoryId, tagIds)
    }
}
